File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/subarray_sum_equals_K/SubarraySumEqualsK.java
package main.java.com.jiang.leetcode100.subarray_sum_equals_K;

import java.util.Scanner;

class Solution {
    public int subarraySum(int[] nums, int k) {
        for(int i = 0; i < nums.length; i++) helper(nums, i, 0, k);
        return result;
    }

    private int result = 0;
    private void helper(int[] nums, int index, int sum, int k) {
        if(index == nums.length) return;
        if(sum + nums[index] == k) result++;
        helper(nums, index + 1, sum + nums[index], k);
    }
}

public class SubarraySumEqualsK {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int k = scanner.nextInt();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution().subarraySum(nums, k));
    }
}

File: Leetcode100/leetcode100_python/SubarraySumEqualsK.py
from typing import List

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefixSum = {0: 1}
        result = 0
        cur_sum = 0
        for num in nums:
            cur_sum += num
            result += prefixSum.get(cur_sum - k, 0)
            prefixSum[cur_sum] = prefixSum.get(cur_sum, 0) + 1
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    k = int(input())
    solution = Solution()
    print(solution.subarraySum(nums, k))